Preserve coalesce state in sparse COO tensor serialization

pearu opened the pull request:
Fixes #101186

Also, resolves the "serialization to preserve coalesced-ness" part in https://github.com/pytorch/pytorch/issues/73479

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102647



cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @mruberry @mikaylagawarecki

pearu committed:
Support sparse COO serialization with coalesce state.

[ghstack-poisoned]

mikaylagawarecki reviewed (commented):
Looks good, thanks!

Could we turn on the [`is_coalesced` checking](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_utils.py#L3019) on [the `self.assertEqual` in this test](https://github.com/pytorch/pytorch/blob/main/test/test_serialization.py#L320) where we test serialization of a sparse_coo tensor? 

As an aside and more for my own understanding I am also curious whether `is_coalesced` is something that should be validated in  `_validate_sparse_coo_tensor_args` which is called [by the `torch.utils._validate_loaded_sparse_tensors` call here](https://github.com/pytorch/pytorch/blob/90b1b17c9f6ab06433fff9484991c5bb40c0763d/torch/serialization.py#L1258)

pearu committed:
Update on "Preserve coalesce state in sparse COO tensor serialization"

Fixes #101186

Also, resolves the "serialization to preserve coalesced-ness" part in https://github.com/pytorch/pytorch/issues/73479




cc alexsamardzic nikitaved cpuhrsch amjames bhosmer mruberry mikaylagawarecki

[ghstack-poisoned]

pearu commented:
> Looks good, thanks!
> 
> Could we turn on the [`is_coalesced` checking](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_utils.py#L3019) on [the `self.assertEqual` in this test](https://github.com/pytorch/pytorch/blob/main/test/test_serialization.py#L320) where we test serialization of a sparse_coo tensor?

Yes, done. Although, it will not be exactly an all-embracing test because this particular test acts on an all-zeros tensor that has the coalesce state always True.

> As an aside and more for my own understanding I am also curious whether `is_coalesced` is something that should be validated in `_validate_sparse_coo_tensor_args` which is called [by the `torch.utils._validate_loaded_sparse_tensors` call here](https://github.com/pytorch/pytorch/blob/90b1b17c9f6ab06433fff9484991c5bb40c0763d/torch/serialization.py#L1258)

I think this question is a more general one: should the `_validate_sparse_coo_tensor_args` function check if COO tensor arguments represent a coalesced COO tensor?
To my opinion, the answer is "No" because the validation function is meant to check if the arguments to `tensor.sparse_coo_tensor` are valid (read: all the layout-specific invariants are satisfied). There is no argument to `tensor.sparse_coo_tensor` that sets the coalesce state.

In addition, I'd say that the "coalesce" property of a COO tensor is data-specific and not really related to the COO layout (saying that a COO tensor is coalesced or not, is similar to saying that a COO tensor represents a triangular tensor or not, for instance). Various algorithms on COO tensors have optimized paths when the COO tensors are coalesced. An uncoalesced tensor is still a valid COO tensor in terms of COO layout invariants, just the algorithms on such tensors are more complex.

Btw, one could ask if there is a way to check if the coalesce state is set properly for a given COO tensor. Currently, we don't have a public method for this. One possible implementation to check if a COO tensor is truly coalesced is as follows:
```python
def is_coalesced(x):
    s = x.is_coalesced()
    y = x._coalesced_(False).coalesce()
    x._coalesced_(s)  # restore the coalesce state of the input
    return (x._indices() == y._indices()).all()
```
but using this in the validate function is inappropriate as it involves expensive operations such as calling the `coalesce` method. In addition, there exist [circumstances](https://github.com/pytorch/pytorch/blob/a6a030a8eb708d0ced13fd2419636e1b24a8e668/aten/src/ATen/native/sparse/SparseTensor.cpp#L797-L799) where one may want to set the coalesce state of a COO tensor to True while the tensor is not truly coalesced.

mikaylagawarecki reviewed (approved):
@pearu Thanks for the detailed explanation regarding whether `is_coalesced` should be validated, that makes sense to me

mikaylagawarecki commented:
@pytorchbot merge

cpuhrsch commented:
@pearu - I guess this comes down to whether we consider it a bug if a Tensor that actually is coalesced isn't flagged as such.

Preserving this annotation for the save and load operations could be seen as a bug fix. It could also be seen as a performance enhancement since it preserves that annotation and thus makes subsequent operations faster.

If we always *must* correctly report whether a Tensor is coalesced it can make certain operations (say addition) very expensive if it requires an expensive additional check to ensure the annotation is always correctly set.

If anything I think we'll want to invest into making [`coalesce`](https://pytorch.org/docs/master/generated/torch.Tensor.coalesce.html?highlight=coalesce#torch.Tensor.coalesce) cheaper / fuseable with torch.compile. Then if a particular operation knows that it'll gain say 10x speedup from explicitly checking coalescion it can risk that additional computation for subsequent outsized gain. That is only if there's some fundamental reason why the uncoalesced form can't be fast.

I don't think we'll find a resolution for this right now (nor should it be high priority given the other items we're working on).
